server/lib/service/reporting.dart

```diff
@@ -0,0 +1,125 @@
+  /// report is not found. Returns the metric value as a string.
+  Future<String> _getLineItemMetric(
+      Int64 advertiserId, Int64 lineItemId, String metric) async {
+    const dataRange = 'PREVIOUS_DAY';
+    final metrics = [metric];
+    final groupBys = [
+      'FILTER_LINE_ITEM',
+      'FILTER_DATE',
+      'FILTER_ADVERTISER_CURRENCY'
+    ];
+
+    // Creates the query for the line item.
+    final queryId = await _createLineItemQuery(
+        advertiserId, lineItemId, dataRange, metrics, groupBys);
+
+    // Get the URL of the report of the created query.
+    final url = await _getReportUrl(queryId);
+
+    try {
+      // Downloads the report and converts it into a map.
+      final reportMap = await _reportToMap(url);
+
+      return reportMap[metricToReportColumnName[metric]];
+    } on ArgumentError {
+      throw ArgumentError('Report with ID $queryId contains missing data.');
+    }
+  }
+
+  /// Creates a query for a line item with [advertiserId] and [lineItemId].
+  ///
+  /// The [dataRange] specifies the range of the report. See
+  /// [QueryMetadata.dataRange]. [metrics] is a list of metrics included in the
+  /// report. [groupBys] specifies which filters the data is grouped by.
+  ///
+  /// Throws an [ApiRequestError] if there is an error with the reporting API.
+  /// Returns the ID of the query.
+  Future<String> _createLineItemQuery(Int64 advertiserId, Int64 lineItemId,
+      String dataRange, List<String> metrics, List<String> groupBys) async {
+    final request = Query()
+      ..metadata = (QueryMetadata()
+        ..title = 'LineItemQuery'
+        ..format = 'CSV'
+        ..dataRange = dataRange)
+      ..params = (Parameters()
+        ..metrics = metrics
+        ..groupBys = groupBys
+        ..filters = [
+          (FilterPair()
+            ..type = 'FILTER_ADVERTISER'
+            ..value = advertiserId.toString()),
+          (FilterPair()
+            ..type = 'FILTER_LINE_ITEM'
+            ..value = lineItemId.toString())
+        ]);
+
+    return (await _api.queries.createquery(request)).queryId;
+  }
+
+  /// Gets the URL of the report with [queryId].
+  ///
+  /// Throws an [ApiRequestError] if there is an error with the reporting API.
+  Future<String> _getReportUrl(String queryId) async {
+    final query = await _api.queries.getquery(queryId);
+    return query.metadata.googleCloudStoragePathForLatestReport;
+  }
+
+  /// Extract the values in the report from [reportUrl] into a map.
+  ///
+  /// Throws an ArgumentError if report is not found or any data is missing.
+  Future<Map<String, String>> _reportToMap(String reportUrl) async {
+    // Sends an HTTP GET request to the [reportUrl].
+    final response = await get(reportUrl);
+
+    // Converts the response body which is in CSV format into a list.
+    final rows = const CsvToListConverter()
+        .convert(response.body, eol: '\n', shouldParseNumbers: false);
+
+    // Converts the list into a map of keys and values.
```

Comment:
Can we add some comment to explain what the key and values is? If it is general, maybe give an example.

Reply:
Sure!

Reply:
Have you updated the comment?

Reply:
I added it to the [doc comment](https://github.com/googleinterns/dv360-excel-plugin/blob/b620fadfc330aa5de033551683fe4280a6f12a34/server/lib/service/reporting.dart#L112) for _retrieveReport().